proxy-server/proxy-server.go

```diff
@@ -157,57 +158,105 @@ func setNewCookie(res http.ResponseWriter) error {
 		Name:    "saturn_token",
 		Value:   tokenString,
 		Expires: expirationTime,
+		Path:    "/",
 	})
 
 	return nil
 }
 
-func validateAuthToken(token *jwt.StandardClaims) bool {
-	// token extra validation. its signature is already verified at this point
-	// considerations: -
-	//			check expiration time for this token
-	//          any extra user access limitations
-	return true
+func validateSaturnToken(saturnToken string, key []byte) (*jwt.StandardClaims, error) {
+	claims := &jwt.StandardClaims{}
+
+	// Parse the JWT string and store the result in `claims`.
+	tkn, err := jwt.ParseWithClaims(saturnToken, claims, func(token *jwt.Token) (interface{}, error) {
+		return key, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	if !tkn.Valid {
+		return nil, errTokenInvalid
+	}
+	if len(claims.Issuer) == 0 {
+		return nil, errNoIssuers
+	}
+	if time.Unix(claims.ExpiresAt, 0).Sub(time.Now()) < 0 {
+		return nil, errExpired
+	}
+	return claims, nil
 }
 
-func redirectToFallBack(res http.ResponseWriter, req *http.Request, error int, origURL string) {
+type proxyAuthResponse struct {
+	SaturnToken string `json:"saturn_token"`
+}
 
-	if debug {
-		log.Printf("fallback code: %d %s", error, origURL)
+func authenticate(res http.ResponseWriter, req *http.Request) bool {
+	res.Header().Add("Cache-Control", "no-cache")
+
+	protocol := "https://"
+	if !useHTTPSForSelfRedirect {
+		protocol = "http://"
 	}
-	switch error {
-	case errNoCookie:
-		log.Printf("Fallback: no cookie token presented or bad cookie")
-	case errExpired:
-		log.Printf("Fallback: cookie expired")
-	case errSignatureInvalid:
-		log.Printf("Fallback: Unauthorized attempt to access resource: wrong signature\n")
-	case errTokenInvalid:
-		log.Printf("Fallback: Unauthorized attempt to access resource: invalid token\n")
-	case errGeneric:
-		log.Printf("Fallback: bad request")
+
+	origURL := protocol + req.Host + req.URL.RequestURI()
+	if debug {
+		log.Printf("fallback: %s", origURL)
 	}
 
 	uniqToken, _ := generateCookieSigningKey(40)
 
+	proxyParams := url.Values{}
+	proxyParams.Add("ret_token", string(uniqToken))
+	proxyParams.Add("proxy_auth", "true")
+	proxyAuthUrl := fallbackURL + "?" + proxyParams.Encode()
+
+	// Check for session cookie
+	if pdcSessionCookie, err := req.Cookie("PDC_SESSION"); err == nil {
```

Comment:
How is this cookie ever being passed in? To my understanding, this is a request coming in for `abc-123-whatever.customer.saturnenterprise.io` - the browser won't send a cookie for `app.customer.saturnenterprise.io` to it.

Reply:
That is correct. I updated Atlas to set the domain for the PDC_SESSION cookie to be the base domain shared by atlas and the proxy: https://github.com/saturncloud/saturn/pull/837/files#diff-8fae5696b6d83acd1a782b93d2a15f5cR287

When I was messing with redirect auth, I kept getting super confused because the PDC_SESSION cookie would show up in requests out of nowhere on redirect to the auth route even though it was never set by the proxy. Turned out that basic GET redirects would work because it could find the `app.customer.saturnenterprise.io` PDC_SESSION cookie, but then XHR GET redirects would fail because XHR is very stubborn about sending cookies.

Reply:
changed in https://github.com/saturncloud/images/pull/52/commits/83d7a6be65a18bf3eb9b030fd766d67f9ab5be5b as per discussion on associated PR